{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e4eeb5a-8cd2-45e4-ad42-95f41f279580",
   "metadata": {},
   "source": [
    "This notebook is created to retrieve and parse the functional information associated to 6 organism models about their protein sequences. The final goal is to create a format available to visualize in Tableau (e.g tsv format) and explore them at the functional level of the orthology relationships among inferred transcriptional regulatory interactions.\n",
    "\n",
    "We'll create (see below) three folders to save our results: *InterproResults*, containing the output of interproscan; *InterproParsedXML*, containing the outputs we'll create and; *InterproFullMerge*, the final tables with all the info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5cadb20-35d6-4ca3-ab20-6d71a6783308",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# If folders dont exist, create them\n",
    "\n",
    "if [ ! -d InterproResults ]; then\n",
    "    mkdir InterproResults\n",
    "fi\n",
    "\n",
    "if [ ! -d InterproParsedXML ]; then \n",
    "    mkdir InterproParsedXML\n",
    "fi\n",
    "\n",
    "if [ ! -d InterproFullMerge ]; then\n",
    "    mkdir InterproFullMerge\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71bc7b9c-601e-46c2-a7de-da284f51dfbc",
   "metadata": {},
   "source": [
    "## Preprocessing\n",
    "\n",
    "In order to retrieve information asscociated about the function of proteins/domains of the sequences, interproscan (only available for based-Linux systems) was used for the six organisms. Basically, Interproscan was ran as is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3927fe71-c26b-4e8a-9489-424a66f086c8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "set -e\n",
    "\n",
    "cd InterproResults\n",
    "for fasta in ../genomes/*faa; do\n",
    "\n",
    "        # Uncomment the following line to run interproscan and replace the path with your\n",
    "        # current sets of interproscan\n",
    "        #./../../../../Descargas/Interpro/interproscan-5.60-92.0/interproscan.sh --verbose -i \"${fasta}\" -f xml,tsv -appl PANTHER,Pfam,FunFam,SUPERFAMILY,CDD --goterms\n",
    "        # To give the opportunity to release memory before the following fasta\n",
    "        sleep 5\n",
    "done\n",
    "\n",
    "cd .."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5614012d-dfb1-47d5-8405-791515b42b4f",
   "metadata": {},
   "source": [
    "Five methods were used (PANTHER,Pfam,FunFam,SUPERFAMILY,CDD) and two output formats were asked. A tsv file containing the structure we are interested to and, a xml containing a wealthier description about the matches of the sequences. Each file is named as the fasta file plus the respective extention, for instance, if the file of the protein sequences of _E.coli_ is **GCF_000005845.2_E_coli_K12_genomic.faa**, the interproscan output will be **GCF_000005845.2_E_coli_K12_genomic.faa.tsv** and **GCF_000005845.2_E_coli_K12_genomic.faa.xml**\n",
    "\n",
    "We will cross the information of both files for each genome. An overview of both files are shown in the following cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1fd5fab5-a92e-4ee8-89dd-54361322b27b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "24d58c00-d8e0-43c1-9060-1c7a3e1d46d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_interpro_results = \"InterproResults\"\n",
    "all_files_names_results = os.listdir(path_interpro_results)\n",
    "\n",
    "# Filtering only the files ending with \"faa.tsv\"\n",
    "tsv_files_names_results = [f for f in all_files_names_results if f.endswith(\"faa.tsv\") ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9336365a-2d30-4b56-a5c8-f6b52aae314e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2_E_coli_K12_genomic.faa.tsv',\n",
       " 'GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv',\n",
       " 'GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv',\n",
       " 'GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv',\n",
       " 'GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv',\n",
       " 'GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_files_names_results.sort()\n",
    "tsv_files_names_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0b517897-ba0b-4c1f-b000-2d2f15825933",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:3.40.50.620:FF:000015</td>\n",
       "      <td>NH(3)-dependent NAD(+) synthetase</td>\n",
       "      <td>1</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>SSF52402</td>\n",
       "      <td>Adenine nucleotide alpha hydrolases-like</td>\n",
       "      <td>3</td>\n",
       "      <td>273</td>\n",
       "      <td>1.450000e-86</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02540</td>\n",
       "      <td>NAD synthase</td>\n",
       "      <td>23</td>\n",
       "      <td>265</td>\n",
       "      <td>1.600000e-81</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR022310</td>\n",
       "      <td>NAD/GMP synthase</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                                 1    2            3   \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275          CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      PANTHER   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275       FunFam   \n",
       "3  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275  SUPERFAMILY   \n",
       "4  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275         Pfam   \n",
       "\n",
       "                            4                                            5   \\\n",
       "0                      cd00553                                 NAD_synthase   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "2  G3DSA:3.40.50.620:FF:000015            NH(3)-dependent NAD(+) synthetase   \n",
       "3                     SSF52402     Adenine nucleotide alpha hydrolases-like   \n",
       "4                      PF02540                                 NAD synthase   \n",
       "\n",
       "   6    7             8  9           10         11                 12  \\\n",
       "0  17  262  1.857550e-84  T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1  11  242  7.600000e-32  T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "2   1  274  0.000000e+00  T  08-02-2023          -                  -   \n",
       "3   3  273  1.450000e-86  T  08-02-2023          -                  -   \n",
       "4  23  265  1.600000e-81  T  08-02-2023  IPR022310   NAD/GMP synthase   \n",
       "\n",
       "                                            13  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  \n",
       "2                                          NaN  \n",
       "3                                          NaN  \n",
       "4                                            -  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We'll load the files later, here I just show one of them\n",
    "pd.read_csv(os.path.join(path_interpro_results, \"GCF_000005845.2_E_coli_K12_genomic.faa.tsv\"),\n",
    "            sep=\"\\t\",\n",
    "            header=None,\n",
    "            index_col=None,\n",
    "            nrows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8fb6a228-d53d-489f-a189-79aa89220762",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?><protein-matches xmlns=\"http://www.ebi.ac.uk/interpro/resources/schemas/interproscan5\" interproscan-version=\"5.60-92.0\">\n",
      "  <protein>\n",
      "    <sequence md5=\"ea37cb76b651f90db9b8d4dbf6861145\">MTLQQQIIKALGAKPQINAEEEIRRSVDFLKSYLQTYPFIKSLVLGISGGQDSTLAGKLCQMAINELRLETGNESLQFIAVRLPYGVQADEQDCQDAIAFIQPDRVLTVNIKGAVLASEQALREAGIELSDFVRGNEKARERMKAQYSIAGMTSGVVVGTDHAAEAITGFFTKYGDGGTDINPLYRLNKRQGKQLLAALACPEHLYKKAPTADLEDDRPSLPDEVALGVTYDNIDDYLEGKNVPQQVARTIENWYLKTEHKRRPPITVFDDFWKK</sequence>\n",
      "    <xref id=\"NP_416254.1\" name=\"NP_416254.1&#9;nadE|NAD synthetase, NH3-dependent|Escherichia coli str. K-12 substr. MG1655\"/>\n",
      "    <matches>\n",
      "      <hmmer3-match evalue=\"0.0\" score=\"522.3\">\n",
      "        <signature ac=\"G3DSA:3.40.50.620:FF:000015\" desc=\"NH(3)-dependent NAD(+) synthetase\">\n",
      "          <signature-library-release library=\"FUNFAM\" version=\"4.3.0\"/>\n",
      "        </signature>\n",
      "        <model-ac>3.40.50.620-FF-000015</model-ac>\n",
      "        <locations>\n",
      "          <hmmer3-location env-end=\"274\" env-start=\"1\" post-processed=\"false\" score=\"522.2\" evalue=\"0.0\" hmm-start=\"1\" hmm-end=\"274\" hmm-length=\"274\" hmm-bounds=\"COMPLETE\" start=\"1\" end=\"274\">\n",
      "            <location-fragments>\n",
      "              <hmmer3-location-fragment start=\"1\" end=\"274\" dc-status=\"CONTINUOUS\"/>\n",
      "            </location-fragments>\n",
      "          </hmmer3-location>\n",
      "        </locations>\n",
      "      </hmmer3-match>\n",
      "      <hmmer3-match evalue=\"1.4E-81\" score=\"273.2\">\n",
      "        <signature ac=\"PF02540\" desc=\"NAD synthase\" name=\"NAD_synthase\">\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head -n 20 \"InterproResults/GCF_000005845.2_E_coli_K12_genomic.faa.xml\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f08e090-3729-4d92-b4be-4c47630d052c",
   "metadata": {},
   "source": [
    "We need to transform the XMF file into something more easy to manipulate and cross with the tsv file. As an initial approach, I'll use RegEx to filter the tags of **xreg**;containing the sequence ID, **signature ac**;the id of the method and job used;**entry ac**; general information and **go-xref category**; descriptions of GO terms\n",
    "\n",
    "First, using the GO terms we'll construct a reference table using bash about all descriptions of the proteins and we'll keep that table separated. In theory, each GO term has a unique ID and that is the reason why this should work.\n",
    "\n",
    "In addition, we'll add a value \"NaN\" to group all the terms with no information and also the header."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d69744a1-8c4f-4c1c-8839-82b93e7cf892",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "category            id_go       description_go\n",
      "BIOLOGICAL_PROCESS  GO:0000041  transition metal ion transport\n",
      "BIOLOGICAL_PROCESS  GO:0000103  sulfate assimilation\n",
      "BIOLOGICAL_PROCESS  GO:0000105  histidine biosynthetic process\n",
      "BIOLOGICAL_PROCESS  GO:0000160  phosphorelay signal transduction system\n",
      "BIOLOGICAL_PROCESS  GO:0000162  tryptophan biosynthetic process\n",
      "BIOLOGICAL_PROCESS  GO:0000256  allantoin catabolic process\n",
      "BIOLOGICAL_PROCESS  GO:0000271  polysaccharide biosynthetic process\n",
      "BIOLOGICAL_PROCESS  GO:0000272  polysaccharide catabolic process\n",
      "BIOLOGICAL_PROCESS  GO:0000413  protein peptidyl-prolyl isomerization\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Empty file\n",
    "printf \"\" > GO_terms_tmp.tsv\n",
    "\n",
    "for XML in InterproResults/*xml; do\n",
    "\n",
    "    # Retrieve go terms; remove attribute names; remove leading spaces and \n",
    "    # replace the first two blank spaces by tabs\n",
    "    PRE_CLEANED_GO=$(grep -E \"<go-xref\\s+category\" \"${XML}\" |\n",
    "                         sed -r 's/(<go-xref category=|db=\"GO\"\\s+id=|name=|\\/>|\")//g' |\n",
    "                         sed -r 's/^\\s+//g' |\n",
    "                         sed -r 's/\\s+/\\t/1' |\n",
    "                         sed -r 's/\\s+/\\t/2')\n",
    "    \n",
    "    echo \"${PRE_CLEANED_GO}\" >> GO_terms_tmp.tsv\n",
    "\n",
    "done\n",
    "\n",
    "# Removing go terms repeated\n",
    "sort GO_terms_tmp.tsv | uniq > GO_terms.tsv && rm GO_terms_tmp.tsv\n",
    "\n",
    "# Adding NaN at the end and headers at the beginning\n",
    "printf \"NaN\\tNaN\\tNaN\\n\" >> GO_terms.tsv\n",
    "sed -i \"1s/^/category\\tid_go\\tdescription_go\\n/\" GO_terms.tsv\n",
    "\n",
    "column -s$'\\t' -t GO_terms.tsv | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6859dda-f868-4c29-a5ff-f61838a88538",
   "metadata": {
    "tags": []
   },
   "source": [
    "As you can see, the new file named *GO_terms.tsv* has three columns: first one with the category of the GO term, a second one with the GO id and the last one with the description about that term.\n",
    "\n",
    "Now it's time to process the remained labels. Because of the complexity of the XML files (e.g many nested tags) the use of libraries such as ```xml.etree.ElementTree``` in pyhon does not seem appropiate in terms of for loops needed and maybe in terms of legibility (but I'll use regex haha), same case for trying to use ```pandas``` so again, I'll do that using bash, feel free to find another way to acomplish this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "436abd88-5288-40f7-8bc9-61cf4b94a4a5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_416254.1     G3DSA:3.40.50.620:FF:000015    NULL_TYPE\n",
      "NP_416254.1     PF02540                        DOMAIN\n",
      "NP_416254.1     PTHR23090                      FAMILY\n",
      "NP_416254.1     cd00553                        FAMILY\n",
      "NP_416254.1     SSF52402                       NULL_TYPE\n",
      "NP_416630.1     PF07694                        DOMAIN\n",
      "NP_416630.1     PF06580                        DOMAIN\n",
      "NP_416630.1     G3DSA:3.30.450.40:FF:000013    NULL_TYPE\n",
      "NP_416630.1     PTHR34220                      NULL_TYPE\n",
      "NP_416630.1     cd16956                        NULL_TYPE\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for XML in InterproResults/*xml; do\n",
    "\n",
    "    OUTPUT_NAME=$(basename \"${XML}\" | sed -r 's/$/\\.tsv/g')\n",
    "\n",
    "    # Retrieve all tags and attributes; retrieve only important attributes; clean attribute names \n",
    "    PRE_CLEANED_XML=$(grep -P \"(<xref |<signature |<entry ac)\" \"${XML}\" |\n",
    "                          grep -Po '(<xref id=\"[^\"]+\"|<signature ac=\"[^\"]+\"|type=\"[^\"]+\")' |\n",
    "                          sed -r 's/^<\\w+ //g')\n",
    "\n",
    "    # Remove quotes \", merge type with method ID into a single row, clean attributes names\n",
    "    PRE_FORMATED_XML=$(echo \"${PRE_CLEANED_XML}\" |\n",
    "                           sed -r 's/\"//g' |\n",
    "                           sed ':r;$!{N;br};s/\\ntype/\\ttype/g' |\n",
    "                           sed -r 's/ac=|type=//g')\n",
    "\n",
    "    # Parser protein ID (clear format) and get tabular form\n",
    "    echo \"${PRE_FORMATED_XML}\" |\n",
    "        perl -ne 'if($_ =~ /id=/){\n",
    "                        chomp($_);\n",
    "                        $header= ($_ =~ s/id=//r);\n",
    "                } else{\n",
    "                        chomp($_);\n",
    "                        $new_string=$header . \"\\t\" . $_;\n",
    "                        $len = 3 - scalar(split(\"\\t\", $new_string));\n",
    "                        print $new_string,\"\\tNULL_TYPE\"x$len,\"\\n\"\n",
    "                }' > InterproParsedXML/$OUTPUT_NAME\n",
    "\n",
    "done\n",
    "\n",
    "# Ecoli file\n",
    "column -s$'\\t' -t InterproParsedXML/GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv | head -n 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fc4bd4d-2daa-4b59-8985-41007dc48b89",
   "metadata": {},
   "source": [
    "From this file, we can load and map by protein ID and Method ID the XML file parsed and the tsv from interproscan for every organism, we'll take _E.coli_ as an example of the logic to follow:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dd62775f-07e6-45fd-85e1-8d60ed6bb8f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Reading two into two dataFrames the tsv file from interproscan and the tsv file obtained from the \n",
    "# processing of the xml file\n",
    "\n",
    "path_interpro_parsed = \"InterproParsedXML/\"\n",
    "\n",
    "xml_parsed_df = pd.read_csv(os.path.join(path_interpro_parsed,\"GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv\"),\n",
    "                            header=None,\n",
    "                            sep=\"\\t\")\n",
    "\n",
    "tsv_df = pd.read_csv(os.path.join(path_interpro_results, \"GCF_000005845.2_E_coli_K12_genomic.faa.tsv\"),\n",
    "                     header=None,\n",
    "                     sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "96e544bc-8538-41a2-ab1c-4df202d32ecf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Creating an index based on protein ID (column 0) and method ID (column 4).\n",
    "# This index will be shared between the two dataframes to cross the information later\n",
    "\n",
    "tsv_df[\"mul_index\"] = tsv_df[0] + \"\\t\" + tsv_df[4]\n",
    "\n",
    "# Replacing NULL values (score -) by np.NaN\n",
    "for col in tsv_df:\n",
    "    tsv_df[col] = tsv_df[col].replace(\"-\", np.NaN)\n",
    "\n",
    "# Replacing the NULL values (\"NULL_TYPE\") by np.NaN\n",
    "xml_parsed_df[2] = xml_parsed_df[2].replace(\"NULL_TYPE\", np.NaN)\n",
    "xml_parsed_df[\"mul_index\"] = xml_parsed_df[0] + \"\\t\" + xml_parsed_df[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c97f5ccc-63a7-4f21-96b5-96a6adc66274",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>mul_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>G3DSA:3.40.50.620:FF:000015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_416254.1\\tG3DSA:3.40.50.620:FF:000015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>PF02540</td>\n",
       "      <td>DOMAIN</td>\n",
       "      <td>NP_416254.1\\tPF02540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>NP_416254.1\\tPTHR23090</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0                            1       2  \\\n",
       "0  NP_416254.1  G3DSA:3.40.50.620:FF:000015     NaN   \n",
       "1  NP_416254.1                      PF02540  DOMAIN   \n",
       "2  NP_416254.1                    PTHR23090  FAMILY   \n",
       "\n",
       "                                  mul_index  \n",
       "0  NP_416254.1\\tG3DSA:3.40.50.620:FF:000015  \n",
       "1                      NP_416254.1\\tPF02540  \n",
       "2                    NP_416254.1\\tPTHR23090  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xml_parsed_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "54325444-0c58-42dc-a474-51b9af1bb84d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>mul_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>NP_416254.1\\tcd00553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>NP_416254.1\\tPTHR23090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:3.40.50.620:FF:000015</td>\n",
       "      <td>NH(3)-dependent NAD(+) synthetase</td>\n",
       "      <td>1</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_416254.1\\tG3DSA:3.40.50.620:FF:000015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0                                 1    2        3  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275  PANTHER   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275   FunFam   \n",
       "\n",
       "                             4                                            5  \\\n",
       "0                      cd00553                                 NAD_synthase   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "2  G3DSA:3.40.50.620:FF:000015            NH(3)-dependent NAD(+) synthetase   \n",
       "\n",
       "    6    7             8  9          10         11                 12  \\\n",
       "0  17  262  1.857550e-84  T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1  11  242  7.600000e-32  T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "2   1  274  0.000000e+00  T  08-02-2023        NaN                NaN   \n",
       "\n",
       "                                            13  \\\n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435   \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435   \n",
       "2                                          NaN   \n",
       "\n",
       "                                  mul_index  \n",
       "0                      NP_416254.1\\tcd00553  \n",
       "1                    NP_416254.1\\tPTHR23090  \n",
       "2  NP_416254.1\\tG3DSA:3.40.50.620:FF:000015  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a0738a78-2b26-4fa8-89d9-1b4328b822a1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>2_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:3.40.50.620:FF:000015</td>\n",
       "      <td>NH(3)-dependent NAD(+) synthetase</td>\n",
       "      <td>1</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0_x                               1_x  2_x        3  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275  PANTHER   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275   FunFam   \n",
       "\n",
       "                             4                                            5  \\\n",
       "0                      cd00553                                 NAD_synthase   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "2  G3DSA:3.40.50.620:FF:000015            NH(3)-dependent NAD(+) synthetase   \n",
       "\n",
       "    6    7             8  9          10         11                 12  \\\n",
       "0  17  262  1.857550e-84  T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1  11  242  7.600000e-32  T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "2   1  274  0.000000e+00  T  08-02-2023        NaN                NaN   \n",
       "\n",
       "                                            13     2_y  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  \n",
       "2                                          NaN     NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Left Join by mul_index\n",
    "full_tsv = tsv_df.merge(xml_parsed_df, on=\"mul_index\", how=\"left\")\n",
    "full_tsv.drop([\"mul_index\",\"0_y\",\"1_y\"], axis=1, inplace=True)\n",
    "full_tsv.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b62169-e45d-4403-a5eb-ce7225e576d0",
   "metadata": {},
   "source": [
    "In the *GO* column we have multiples values in a single row, in order to connect the table we are creating with the **GO_terms.tsv** table, we'll convert the DataFrame in a \"long\" format dataframe expanding the multiples values found in *GO* to multiples rows describing the same result but, this time, each row will have only one GO term as exemplified below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c2e9e052-e706-4f55-992b-41762f40b0bc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GO:1|GO:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GO:3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ids         go\n",
       "0    0  GO:1|GO:2\n",
       "1    1        NaN\n",
       "2    2       GO:3"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids = np.arange(0,3,1)\n",
    "go_terms = [\"GO:1|GO:2\",np.NaN,\"GO:3\"]\n",
    "\n",
    "df_shown = pd.DataFrame({\"ids\":ids, \"go\":go_terms})\n",
    "df_shown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "52bfd637-58e4-4562-b411-ae7ddcd8184c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GO:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>GO:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>GO:3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ids    go\n",
       "0    0  GO:1\n",
       "1    0  GO:2\n",
       "2    1   NaN\n",
       "3    2  GO:3"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids = np.array([0,0,1,2])\n",
    "go_terms = [\"GO:1\",\"GO:2\", np.NaN, \"GO:3\"]\n",
    "\n",
    "df_expected = pd.DataFrame({\"ids\":ids, \"go\":go_terms})\n",
    "df_expected"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c42b99-8a5c-40fc-a6f6-f8cff763ef16",
   "metadata": {},
   "source": [
    "Note that maintaining the index is important since it will be used to map between this new format and the interproscan table format we want in the end."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "937e65fd-d5a4-419c-b6f4-6c2358771cdd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>2_y</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0004359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0_x                               1_x  2_x        3          4  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD    cd00553   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD    cd00553   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD    cd00553   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD    cd00553   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275  PANTHER  PTHR23090   \n",
       "\n",
       "                                             5   6    7             8  9  \\\n",
       "0                                 NAD_synthase  17  262  1.857550e-84  T   \n",
       "0                                 NAD_synthase  17  262  1.857550e-84  T   \n",
       "0                                 NAD_synthase  17  262  1.857550e-84  T   \n",
       "0                                 NAD_synthase  17  262  1.857550e-84  T   \n",
       "1  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE  11  242  7.600000e-32  T   \n",
       "\n",
       "           10         11                 12  \\\n",
       "0  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "0  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "0  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "0  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "\n",
       "                                            13     2_y       value  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0003952  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0004359  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0005737  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0009435  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0003952  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert the GO terms found into a dataframe of multiples columns depending of the number of GO terms\n",
    "full_go_terms_df = full_tsv[13].str.split(\"|\").apply(pd.Series)\n",
    "\n",
    "# \"Merge\" all the columns from the same index into multiples rows of one colum\n",
    "full_go_terms_merged_df = full_go_terms_df.copy().T\n",
    "full_go_terms_merged_df = full_go_terms_merged_df.melt(var_name=\"index_tmp\").drop_duplicates()\n",
    "\n",
    "# Masking to remove duplicated index with np.NaN value\n",
    "mask = ~(full_go_terms_merged_df[\"index_tmp\"].duplicated(keep=False) & full_go_terms_merged_df[\"value\"].isna())\n",
    "full_go_terms_merged_df = full_go_terms_merged_df[mask].copy().set_index(\"index_tmp\")\n",
    "\n",
    "# Left join with the interproscan table we have been processing in the last cells\n",
    "full_df = full_tsv.merge(full_go_terms_merged_df, how=\"left\", right_index=True, left_index=True).copy()\n",
    "full_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "42b73d18-8b85-4134-9bc8-a3efd9d88a4f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>2_y</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0004359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21627</th>\n",
       "      <td>NP_416130.3</td>\n",
       "      <td>db79e4dbecd21e0c3a7736850a23e810</td>\n",
       "      <td>391</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:2.60.120.10:FF:000076</td>\n",
       "      <td>Mannose-6-phosphate isomerase</td>\n",
       "      <td>301</td>\n",
       "      <td>382</td>\n",
       "      <td>1.400000e-53</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21628</th>\n",
       "      <td>NP_416130.3</td>\n",
       "      <td>db79e4dbecd21e0c3a7736850a23e810</td>\n",
       "      <td>391</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF01238</td>\n",
       "      <td>Phosphomannose isomerase type I C-terminal</td>\n",
       "      <td>310</td>\n",
       "      <td>357</td>\n",
       "      <td>9.800000e-16</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR046456</td>\n",
       "      <td>Phosphomannose isomerase type I, C-terminal do...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DOMAIN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21629</th>\n",
       "      <td>NP_416130.3</td>\n",
       "      <td>db79e4dbecd21e0c3a7736850a23e810</td>\n",
       "      <td>391</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF20511</td>\n",
       "      <td>Phosphomannose isomerase type I, catalytic domain</td>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>2.300000e-54</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR046457</td>\n",
       "      <td>Phosphomannose isomerase type I, catalytic domain</td>\n",
       "      <td>GO:0004476|GO:0008270</td>\n",
       "      <td>DOMAIN</td>\n",
       "      <td>GO:0004476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21629</th>\n",
       "      <td>NP_416130.3</td>\n",
       "      <td>db79e4dbecd21e0c3a7736850a23e810</td>\n",
       "      <td>391</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF20511</td>\n",
       "      <td>Phosphomannose isomerase type I, catalytic domain</td>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "      <td>2.300000e-54</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR046457</td>\n",
       "      <td>Phosphomannose isomerase type I, catalytic domain</td>\n",
       "      <td>GO:0004476|GO:0008270</td>\n",
       "      <td>DOMAIN</td>\n",
       "      <td>GO:0008270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21630</th>\n",
       "      <td>NP_416130.3</td>\n",
       "      <td>db79e4dbecd21e0c3a7736850a23e810</td>\n",
       "      <td>391</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF20512</td>\n",
       "      <td>Phosphomannose isomerase type I, helical inser...</td>\n",
       "      <td>157</td>\n",
       "      <td>236</td>\n",
       "      <td>1.200000e-16</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR046458</td>\n",
       "      <td>Phosphomannose isomerase type I, helical inser...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DOMAIN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26716 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0_x                               1_x  2_x        3  \\\n",
       "0      NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD   \n",
       "0      NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD   \n",
       "0      NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD   \n",
       "0      NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275      CDD   \n",
       "1      NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145  275  PANTHER   \n",
       "...            ...                               ...  ...      ...   \n",
       "21627  NP_416130.3  db79e4dbecd21e0c3a7736850a23e810  391   FunFam   \n",
       "21628  NP_416130.3  db79e4dbecd21e0c3a7736850a23e810  391     Pfam   \n",
       "21629  NP_416130.3  db79e4dbecd21e0c3a7736850a23e810  391     Pfam   \n",
       "21629  NP_416130.3  db79e4dbecd21e0c3a7736850a23e810  391     Pfam   \n",
       "21630  NP_416130.3  db79e4dbecd21e0c3a7736850a23e810  391     Pfam   \n",
       "\n",
       "                                 4  \\\n",
       "0                          cd00553   \n",
       "0                          cd00553   \n",
       "0                          cd00553   \n",
       "0                          cd00553   \n",
       "1                        PTHR23090   \n",
       "...                            ...   \n",
       "21627  G3DSA:2.60.120.10:FF:000076   \n",
       "21628                      PF01238   \n",
       "21629                      PF20511   \n",
       "21629                      PF20511   \n",
       "21630                      PF20512   \n",
       "\n",
       "                                                       5    6    7  \\\n",
       "0                                           NAD_synthase   17  262   \n",
       "0                                           NAD_synthase   17  262   \n",
       "0                                           NAD_synthase   17  262   \n",
       "0                                           NAD_synthase   17  262   \n",
       "1            NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   11  242   \n",
       "...                                                  ...  ...  ...   \n",
       "21627                      Mannose-6-phosphate isomerase  301  382   \n",
       "21628         Phosphomannose isomerase type I C-terminal  310  357   \n",
       "21629  Phosphomannose isomerase type I, catalytic domain    1  150   \n",
       "21629  Phosphomannose isomerase type I, catalytic domain    1  150   \n",
       "21630  Phosphomannose isomerase type I, helical inser...  157  236   \n",
       "\n",
       "                  8  9          10         11  \\\n",
       "0      1.857550e-84  T  08-02-2023  IPR003694   \n",
       "0      1.857550e-84  T  08-02-2023  IPR003694   \n",
       "0      1.857550e-84  T  08-02-2023  IPR003694   \n",
       "0      1.857550e-84  T  08-02-2023  IPR003694   \n",
       "1      7.600000e-32  T  08-02-2023  IPR003694   \n",
       "...             ... ..         ...        ...   \n",
       "21627  1.400000e-53  T  08-02-2023        NaN   \n",
       "21628  9.800000e-16  T  08-02-2023  IPR046456   \n",
       "21629  2.300000e-54  T  08-02-2023  IPR046457   \n",
       "21629  2.300000e-54  T  08-02-2023  IPR046457   \n",
       "21630  1.200000e-16  T  08-02-2023  IPR046458   \n",
       "\n",
       "                                                      12  \\\n",
       "0                                      NAD(+) synthetase   \n",
       "0                                      NAD(+) synthetase   \n",
       "0                                      NAD(+) synthetase   \n",
       "0                                      NAD(+) synthetase   \n",
       "1                                      NAD(+) synthetase   \n",
       "...                                                  ...   \n",
       "21627                                                NaN   \n",
       "21628  Phosphomannose isomerase type I, C-terminal do...   \n",
       "21629  Phosphomannose isomerase type I, catalytic domain   \n",
       "21629  Phosphomannose isomerase type I, catalytic domain   \n",
       "21630  Phosphomannose isomerase type I, helical inser...   \n",
       "\n",
       "                                                13     2_y       value  \n",
       "0      GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0003952  \n",
       "0      GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0004359  \n",
       "0      GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0005737  \n",
       "0      GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0009435  \n",
       "1      GO:0003952|GO:0004359|GO:0005737|GO:0009435  FAMILY  GO:0003952  \n",
       "...                                            ...     ...         ...  \n",
       "21627                                          NaN     NaN         NaN  \n",
       "21628                                          NaN  DOMAIN         NaN  \n",
       "21629                        GO:0004476|GO:0008270  DOMAIN  GO:0004476  \n",
       "21629                        GO:0004476|GO:0008270  DOMAIN  GO:0008270  \n",
       "21630                                          NaN  DOMAIN         NaN  \n",
       "\n",
       "[26716 rows x 16 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "66a314f4-7efe-4a60-89fb-80138a9100e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>md5</th>\n",
       "      <th>protein_len</th>\n",
       "      <th>method</th>\n",
       "      <th>id_method</th>\n",
       "      <th>description</th>\n",
       "      <th>start_loc</th>\n",
       "      <th>end_loc</th>\n",
       "      <th>score</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>interpro_access</th>\n",
       "      <th>interpro_description</th>\n",
       "      <th>type</th>\n",
       "      <th>GO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0004359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>FAMILY</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id                               md5  protein_len   method  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145          275      CDD   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145          275      CDD   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145          275      CDD   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145          275      CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145          275  PANTHER   \n",
       "\n",
       "   id_method                                  description  start_loc  end_loc  \\\n",
       "0    cd00553                                 NAD_synthase         17      262   \n",
       "0    cd00553                                 NAD_synthase         17      262   \n",
       "0    cd00553                                 NAD_synthase         17      262   \n",
       "0    cd00553                                 NAD_synthase         17      262   \n",
       "1  PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE         11      242   \n",
       "\n",
       "          score status        date interpro_access interpro_description  \\\n",
       "0  1.857550e-84      T  08-02-2023       IPR003694    NAD(+) synthetase   \n",
       "0  1.857550e-84      T  08-02-2023       IPR003694    NAD(+) synthetase   \n",
       "0  1.857550e-84      T  08-02-2023       IPR003694    NAD(+) synthetase   \n",
       "0  1.857550e-84      T  08-02-2023       IPR003694    NAD(+) synthetase   \n",
       "1  7.600000e-32      T  08-02-2023       IPR003694    NAD(+) synthetase   \n",
       "\n",
       "     type          GO  \n",
       "0  FAMILY  GO:0003952  \n",
       "0  FAMILY  GO:0004359  \n",
       "0  FAMILY  GO:0005737  \n",
       "0  FAMILY  GO:0009435  \n",
       "1  FAMILY  GO:0003952  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now lets remove the \"GO\" column and rename the columns according to the doc of interproscan\n",
    "\n",
    "full_df.drop(13, axis=1, inplace=True)\n",
    "\n",
    "full_df.rename({\"0_x\": \"id\",\n",
    "                 \"1_x\":\"md5\",\n",
    "                 \"2_x\": \"protein_len\",\n",
    "                 3: \"method\", \n",
    "                 4:\"id_method\", \n",
    "                 5:\"description\", \n",
    "                 6:\"start_loc\", \n",
    "                 7:\"end_loc\", \n",
    "                 8:\"score\", \n",
    "                 9:\"status\", \n",
    "                 10:\"date\", \n",
    "                 11:\"interpro_access\", \n",
    "                 12:\"interpro_description\", \n",
    "                 \"2_y\":\"type\", \n",
    "                 \"value\":\"GO\"}, axis=1, inplace=True)\n",
    "\n",
    "full_df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74694ea8-50f2-458b-8753-4aef3a1d5706",
   "metadata": {},
   "source": [
    "Now we need to do the same for every organism, I'll create some functions of the codes shown above, so they are going to do all the hard work. In addition, a short code will be added to save into a file the tables created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "62332ac0-21fd-4ec8-8df7-4fcff5537067",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def LoadFiles(xml_file_name:str, tsv_file_name:str, keydict:dict=None) -> (pd.DataFrame, pd.DataFrame):\n",
    "    \n",
    "    # This function read a tsv file and a parsed xml file obtained from interproscan into a pandas DataFrame\n",
    "    \n",
    "    print(f\"Reading files into DataFrames: {xml_file_name}\")\n",
    "    print(f\"                               {tsv_file_name}\")\n",
    "    \n",
    "    xml_df = pd.read_csv(xml_file_name, **keydict)\n",
    "    tsv_df = pd.read_csv(tsv_file_name, **keydict)\n",
    "    \n",
    "    return xml_df, tsv_df\n",
    "    \n",
    "def ParseFrames(xml_df:pd.DataFrame, tsv_df:pd.DataFrame, shared_index:str) -> (pd.DataFrame, pd.DataFrame):\n",
    "    \n",
    "    # This function create a shared index between two data frames and replace null values by np.NaN\n",
    "    \n",
    "    print(f\"Parsing frames using the shared_index = {shared_index}\")\n",
    "    \n",
    "    tsv_df[shared_index] = tsv_df[0] + \"\\t\" + tsv_df[4]\n",
    "\n",
    "    # Replacing NULL values (score -) by NaN\n",
    "    for col in tsv_df:\n",
    "        tsv_df[col] = tsv_df[col].replace(\"-\", np.NaN)\n",
    "\n",
    "    xml_df[2] = xml_df[2].replace(\"NULL_TYPE\", np.NaN)\n",
    "    xml_df[shared_index] = xml_df[0] + \"\\t\" + xml_df[1]\n",
    "    \n",
    "    return xml_df, tsv_df\n",
    "\n",
    "def JoinFrames(left:pd.DataFrame, right:pd.DataFrame, keydict:dict, drop:list=None) -> pd.DataFrame:\n",
    "    \n",
    "    # This function do a join operation between two dataframes\n",
    "    \n",
    "    print(f\"Joining frames with parameters: {keydict}\")\n",
    "    df = left.merge(right, **keydict)\n",
    "    \n",
    "    if drop:\n",
    "        df.drop(drop, axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def ParseGo(data:pd.DataFrame, col_target, sep_target, keydict:dict) -> pd.DataFrame:\n",
    "    \n",
    "    # This function merge the GO term found in a single-line format in a dataframe \n",
    "    # into multiples lines containing only one GO term in a dataframe  \n",
    "    \n",
    "    print(f\"Parsing GO terms with parameters {keydict}\")\n",
    "    \n",
    "    data_go = data[col_target].str.split(sep_target).apply(pd.Series)\n",
    "\n",
    "    # \"Merge\" all the columns from the same index into multiples rows of one colum\n",
    "    data_go_merged = data_go.copy().T\n",
    "    data_go_merged = data_go_merged.melt(var_name=\"index_tmp\").drop_duplicates()\n",
    "\n",
    "    # Masking to remove duplicated index with np.NaN value\n",
    "    mask = ~(data_go_merged[\"index_tmp\"].duplicated(keep=False) & data_go_merged[\"value\"].isna())\n",
    "    data_go_merged = data_go_merged[mask].copy().set_index(\"index_tmp\")\n",
    "\n",
    "    # Left join with the interproscan table we have been processing in the last cells\n",
    "    full_df = data.merge(data_go_merged, **keydict).copy()\n",
    "    \n",
    "    return full_df.drop(col_target, axis=1)\n",
    "    \n",
    "def RenameFrame(DataFrame:pd.DataFrame) -> pd.DataFrame:\n",
    "    \n",
    "    # This function rename the columns of a dataframe according to the interpro documentation\n",
    "    print(\"Renaming frame\")\n",
    "    \n",
    "    DataFrame.rename({\"0_x\": \"id\",\n",
    "                 \"1_x\":\"md5\",\n",
    "                 \"2_x\": \"protein_len\",\n",
    "                 3: \"method\", \n",
    "                 4:\"id_method\", \n",
    "                 5:\"description\", \n",
    "                 6:\"start_loc\", \n",
    "                 7:\"end_loc\", \n",
    "                 8:\"score\", \n",
    "                 9:\"status\", \n",
    "                 10:\"date\", \n",
    "                 11:\"interpro_access\", \n",
    "                 12:\"interpro_description\", \n",
    "                 \"2_y\":\"type\", \n",
    "                 \"value\":\"GO\"}, axis=1, inplace=True)\n",
    "    \n",
    "    return DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6442ecc-f295-4097-844d-a1ab4e370458",
   "metadata": {},
   "source": [
    "TSV file are already loaded, we'll need to load now the new created xml.tsv files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fcc1464a-4dc5-4dbf-b1df-424bacd07e36",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_files_names_parsed = os.listdir(path_interpro_parsed)\n",
    "\n",
    "# Filter only the files ending with \"xml.tsv\"\n",
    "xml_tsv_files_names = [f for f in all_files_names_parsed if f.endswith(\"xml.tsv\") ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b6952770-8ae0-45b2-979c-9ed4a1e17903",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv',\n",
       " 'GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.xml.tsv',\n",
       " 'GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.xml.tsv',\n",
       " 'GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.xml.tsv',\n",
       " 'GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.xml.tsv',\n",
       " 'GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.xml.tsv']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xml_tsv_files_names.sort()\n",
    "xml_tsv_files_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "83a97453-005e-435c-98ee-bb1eaf5d4848",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2_E_coli_K12_genomic.faa.tsv',\n",
       " 'GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv',\n",
       " 'GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv',\n",
       " 'GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv',\n",
       " 'GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv',\n",
       " 'GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_files_names_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e4621950-105e-4e90-8d76-552639b03d3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading files into DataFrames: InterproParsedXML/GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000005845.2_E_coli_K12_genomic.faa.tsv\n",
      "Parsing frames using the shared_index = mul_index\n",
      "Joining frames with parameters: {'on': 'mul_index', 'how': 'left'}\n",
      "Parsing GO terms with parameters {'how': 'left', 'right_index': True, 'left_index': True}\n",
      "Renaming frame\n",
      "Saving frame into InterproFullMerge/GCF_000005845.2_E_coli_K12_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv\n",
      "Parsing frames using the shared_index = mul_index\n",
      "Joining frames with parameters: {'on': 'mul_index', 'how': 'left'}\n",
      "Parsing GO terms with parameters {'how': 'left', 'right_index': True, 'left_index': True}\n",
      "Renaming frame\n",
      "Saving frame into InterproFullMerge/GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv\n",
      "Parsing frames using the shared_index = mul_index\n",
      "Joining frames with parameters: {'on': 'mul_index', 'how': 'left'}\n",
      "Parsing GO terms with parameters {'how': 'left', 'right_index': True, 'left_index': True}\n",
      "Renaming frame\n",
      "Saving frame into InterproFullMerge/GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv\n",
      "Parsing frames using the shared_index = mul_index\n",
      "Joining frames with parameters: {'on': 'mul_index', 'how': 'left'}\n",
      "Parsing GO terms with parameters {'how': 'left', 'right_index': True, 'left_index': True}\n",
      "Renaming frame\n",
      "Saving frame into InterproFullMerge/GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv\n",
      "Parsing frames using the shared_index = mul_index\n",
      "Joining frames with parameters: {'on': 'mul_index', 'how': 'left'}\n",
      "Parsing GO terms with parameters {'how': 'left', 'right_index': True, 'left_index': True}\n",
      "Renaming frame\n",
      "Saving frame into InterproFullMerge/GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv\n",
      "Parsing frames using the shared_index = mul_index\n",
      "Joining frames with parameters: {'on': 'mul_index', 'how': 'left'}\n",
      "Parsing GO terms with parameters {'how': 'left', 'right_index': True, 'left_index': True}\n",
      "Renaming frame\n",
      "Saving frame into InterproFullMerge/GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv.full.tsv\n",
      "\n",
      "All done\n"
     ]
    }
   ],
   "source": [
    "# Arguments for the functions and folder to save files\n",
    "output_dir = \"InterproFullMerge\"\n",
    "arguments_load_function = {\"header\": None, \"sep\":\"\\t\"}\n",
    "arguments_join_function = {\"on\":\"mul_index\", \"how\":\"left\"}\n",
    "arguments_parse_go_function = {\"how\":\"left\", \"right_index\":True, \"left_index\":True}\n",
    "\n",
    "# Doing the same as shown above but using all organisms\n",
    "for t_f, x_f in zip(tsv_files_names_results, xml_tsv_files_names):\n",
    "    \n",
    "    tsv_path = os.path.join(path_interpro_results, t_f)\n",
    "    xml_path = os.path.join(path_interpro_parsed, x_f)\n",
    "    \n",
    "    xml_df, tsv_df = LoadFiles(xml_path, tsv_path, arguments_load_function)\n",
    "    xml_df, tsv_df = ParseFrames(xml_df, tsv_df, \"mul_index\")\n",
    "    \n",
    "    full_df = JoinFrames(tsv_df, xml_df, arguments_join_function, [\"mul_index\",\"0_y\",\"1_y\"])\n",
    "    full_df = ParseGo(full_df, 13, \"|\", arguments_parse_go_function)\n",
    "    \n",
    "    full_df = RenameFrame(full_df)\n",
    "    \n",
    "    # Saving frame into a file ending with \".full\" \n",
    "    output_file_path = os.path.join(output_dir, t_f + \".full.tsv\")\n",
    "    print(f\"Saving frame into {output_file_path}\\n\")\n",
    "    full_df.to_csv(output_file_path, sep=\"\\t\", header=True, index=True, na_rep=\"NaN\", index_label=\"id_row\")\n",
    "\n",
    "print(\"All done\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
