{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e4eeb5a-8cd2-45e4-ad42-95f41f279580",
   "metadata": {},
   "source": [
    "This notebook is created to retrieve and parse the functional information associated to 6 organism models about their protein sequences. The final goal is to create a format available to visualize in Tableau (e.g tsv format) and explore them at the functional level of the orthology relationships among inferred transcriptional regulatory interactions.\n",
    "\n",
    "We'll create (see below) three folders to save our results: *InterproResults*, containing the output of interproscan; *InterproParsedXML*, containing the outputs we'll create and; *InterproFullMerge*, the final tables with all the info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5cadb20-35d6-4ca3-ab20-6d71a6783308",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# If folders dont exist, create them\n",
    "\n",
    "if [ ! -d InterproResults ]; then\n",
    "    mkdir InterproResults\n",
    "fi\n",
    "\n",
    "if [ ! -d InterproParsedXML ]; then \n",
    "    mkdir InterproParsedXML\n",
    "fi\n",
    "\n",
    "if [ ! -d InterproFullMerge ]; then\n",
    "    mkdir InterproFullMerge\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7731716-63a1-4106-8dff-44ad14aa81a3",
   "metadata": {},
   "source": [
    "and we'll import the python libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e32148f-abec-4b9d-8fb4-b7631b9fcc4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71bc7b9c-601e-46c2-a7de-da284f51dfbc",
   "metadata": {},
   "source": [
    "## Preprocessing\n",
    "\n",
    "In order to retrieve information asscociated about the function of proteins/domains of the sequences, interproscan (only available for based-Linux systems) was used for the six organisms. Basically, Interproscan was ran as is shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3927fe71-c26b-4e8a-9489-424a66f086c8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd InterproResults\n",
    "for fasta in ../genomes/*faa; do\n",
    "\n",
    "        # Uncomment the following line to run interproscan and replace the path with your\n",
    "        # current sets of interproscan\n",
    "        #./../../../../Descargas/Interpro/interproscan-5.60-92.0/interproscan.sh --verbose -i \"${fasta}\" -f xml,tsv -appl PANTHER,Pfam,FunFam,SUPERFAMILY,CDD --goterms\n",
    "        # To give the opportunity to release memory before the following fasta\n",
    "        sleep 5\n",
    "done\n",
    "\n",
    "cd .."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5614012d-dfb1-47d5-8405-791515b42b4f",
   "metadata": {},
   "source": [
    "Five methods were used (PANTHER,Pfam,FunFam,SUPERFAMILY,CDD) and two output formats were asked. A tsv file containing the structure we are interested to and, a xml containing a wealthier description about the matches of the sequences. Each file is named as the fasta file plus the respective extention, for instance, if the file of the protein sequences of _E.coli_ is **GCF_000005845.2_E_coli_K12_genomic.faa**, the interproscan output will be **GCF_000005845.2_E_coli_K12_genomic.faa.tsv** and **GCF_000005845.2_E_coli_K12_genomic.faa.xml**. \n",
    "\n",
    "We are goin to add some headers to the tsv files. The description of those can be found in the respective documentation of [interproscan](https://interproscan-docs.readthedocs.io/en/latest/OutputFormats.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9ae0ebc2-2204-48c3-967f-b28699bb7c8f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking headers in \n",
      "\tAdding header\n",
      "Checking headers in \n",
      "\tAdding header\n",
      "Checking headers in \n",
      "\tAdding header\n",
      "Checking headers in \n",
      "\tAdding header\n",
      "Checking headers in \n",
      "\tAdding header\n",
      "Checking headers in \n",
      "\tAdding header\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for tsv in InterproResults/*tsv; do\n",
    "    \n",
    "    # Note below we are modifying inplace, therefore, I'll check if the headers are already there\n",
    "    HEADER=$(head -n 1 \"${tsv}\")\n",
    "    EXPECTED_HEADER=$(printf \"id_protein\\tmd5\\tseq_len\\tanalysis\\tid_analysis\\tdesc_analysis\\tstart\\tstop\\tscore\\tstatus\\tdate\\tid_inter\\tdesc_inter\\tgo\\n\")\n",
    "    \n",
    "    echo \"Checking headers in ${net}\"\n",
    "    \n",
    "    if [ \"${HEADER}\" != \"${EXPECTED_HEADER}\" ]; then\n",
    "        printf \"\\tAdding header\\n\"\n",
    "        sed -i \"1s/^/${EXPECTED_HEADER}\\n/\" \"${tsv}\"\n",
    "    fi\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a95721a7-d3fe-4f5c-8f79-43a743aecb68",
   "metadata": {},
   "source": [
    "We will cross the information of both files for each genome later. Now, an overview of both files are shown in the following cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "24d58c00-d8e0-43c1-9060-1c7a3e1d46d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_interpro_results = \"InterproResults\"\n",
    "all_files_names_results = os.listdir(path_interpro_results)\n",
    "\n",
    "# Filtering only the files ending with \"faa.tsv\"\n",
    "tsv_files_names_results = [f for f in all_files_names_results if f.endswith(\"faa.tsv\") ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9336365a-2d30-4b56-a5c8-f6b52aae314e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2_E_coli_K12_genomic.faa.tsv',\n",
       " 'GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv',\n",
       " 'GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv',\n",
       " 'GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv',\n",
       " 'GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv',\n",
       " 'GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_files_names_results.sort()\n",
    "tsv_files_names_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0b517897-ba0b-4c1f-b000-2d2f15825933",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_protein</th>\n",
       "      <th>md5</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>analysis</th>\n",
       "      <th>id_analysis</th>\n",
       "      <th>desc_analysis</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>id_inter</th>\n",
       "      <th>desc_inter</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:3.40.50.620:FF:000015</td>\n",
       "      <td>NH(3)-dependent NAD(+) synthetase</td>\n",
       "      <td>1</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>SSF52402</td>\n",
       "      <td>Adenine nucleotide alpha hydrolases-like</td>\n",
       "      <td>3</td>\n",
       "      <td>273</td>\n",
       "      <td>1.450000e-86</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF02540</td>\n",
       "      <td>NAD synthase</td>\n",
       "      <td>23</td>\n",
       "      <td>265</td>\n",
       "      <td>1.600000e-81</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR022310</td>\n",
       "      <td>NAD/GMP synthase</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id_protein                               md5  seq_len     analysis  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275          CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      PANTHER   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275       FunFam   \n",
       "3  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275  SUPERFAMILY   \n",
       "4  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275         Pfam   \n",
       "\n",
       "                   id_analysis                                desc_analysis  \\\n",
       "0                      cd00553                                 NAD_synthase   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "2  G3DSA:3.40.50.620:FF:000015            NH(3)-dependent NAD(+) synthetase   \n",
       "3                     SSF52402     Adenine nucleotide alpha hydrolases-like   \n",
       "4                      PF02540                                 NAD synthase   \n",
       "\n",
       "   start  stop         score status        date   id_inter         desc_inter  \\\n",
       "0     17   262  1.857550e-84      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1     11   242  7.600000e-32      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "2      1   274  0.000000e+00      T  08-02-2023          -                  -   \n",
       "3      3   273  1.450000e-86      T  08-02-2023          -                  -   \n",
       "4     23   265  1.600000e-81      T  08-02-2023  IPR022310   NAD/GMP synthase   \n",
       "\n",
       "                                            go  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  \n",
       "2                                          NaN  \n",
       "3                                          NaN  \n",
       "4                                            -  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We'll load the files later, here I just show one of them\n",
    "pd.read_csv(os.path.join(path_interpro_results, \"GCF_000005845.2_E_coli_K12_genomic.faa.tsv\"),\n",
    "            sep=\"\\t\",\n",
    "            nrows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8fb6a228-d53d-489f-a189-79aa89220762",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?><protein-matches xmlns=\"http://www.ebi.ac.uk/interpro/resources/schemas/interproscan5\" interproscan-version=\"5.60-92.0\">\n",
      "  <protein>\n",
      "    <sequence md5=\"ea37cb76b651f90db9b8d4dbf6861145\">MTLQQQIIKALGAKPQINAEEEIRRSVDFLKSYLQTYPFIKSLVLGISGGQDSTLAGKLCQMAINELRLETGNESLQFIAVRLPYGVQADEQDCQDAIAFIQPDRVLTVNIKGAVLASEQALREAGIELSDFVRGNEKARERMKAQYSIAGMTSGVVVGTDHAAEAITGFFTKYGDGGTDINPLYRLNKRQGKQLLAALACPEHLYKKAPTADLEDDRPSLPDEVALGVTYDNIDDYLEGKNVPQQVARTIENWYLKTEHKRRPPITVFDDFWKK</sequence>\n",
      "    <xref id=\"NP_416254.1\" name=\"NP_416254.1&#9;nadE|NAD synthetase, NH3-dependent|Escherichia coli str. K-12 substr. MG1655\"/>\n",
      "    <matches>\n",
      "      <hmmer3-match evalue=\"0.0\" score=\"522.3\">\n",
      "        <signature ac=\"G3DSA:3.40.50.620:FF:000015\" desc=\"NH(3)-dependent NAD(+) synthetase\">\n",
      "          <signature-library-release library=\"FUNFAM\" version=\"4.3.0\"/>\n",
      "        </signature>\n",
      "        <model-ac>3.40.50.620-FF-000015</model-ac>\n",
      "        <locations>\n",
      "          <hmmer3-location env-end=\"274\" env-start=\"1\" post-processed=\"false\" score=\"522.2\" evalue=\"0.0\" hmm-start=\"1\" hmm-end=\"274\" hmm-length=\"274\" hmm-bounds=\"COMPLETE\" start=\"1\" end=\"274\">\n",
      "            <location-fragments>\n",
      "              <hmmer3-location-fragment start=\"1\" end=\"274\" dc-status=\"CONTINUOUS\"/>\n",
      "            </location-fragments>\n",
      "          </hmmer3-location>\n",
      "        </locations>\n",
      "      </hmmer3-match>\n",
      "      <hmmer3-match evalue=\"1.4E-81\" score=\"273.2\">\n",
      "        <signature ac=\"PF02540\" desc=\"NAD synthase\" name=\"NAD_synthase\">\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head -n 20 \"InterproResults/GCF_000005845.2_E_coli_K12_genomic.faa.xml\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f08e090-3729-4d92-b4be-4c47630d052c",
   "metadata": {},
   "source": [
    "We need to transform the XMF file into something more easy to manipulate and cross with the tsv file. As an initial approach, I'll use RegEx to filter the tags of **xreg**;containing the sequence ID, **signature ac**;the id of the method and job used;**entry ac**; general information, **protein**; indicates a protein or group of proteins with same results and, **go-xref category**; descriptions of GO terms\n",
    "\n",
    "First, using the GO terms we'll construct a reference table using bash about all descriptions of the proteins and we'll keep that table separated. In theory, each GO term has a unique ID and that is the reason why this should work.\n",
    "\n",
    "In addition, we'll add a value \"NaN\" to group all the terms with no information and also the header."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d69744a1-8c4f-4c1c-8839-82b93e7cf892",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "category            id_go       description_go\n",
      "BIOLOGICAL_PROCESS  GO:0000041  transition metal ion transport\n",
      "BIOLOGICAL_PROCESS  GO:0000103  sulfate assimilation\n",
      "BIOLOGICAL_PROCESS  GO:0000105  histidine biosynthetic process\n",
      "BIOLOGICAL_PROCESS  GO:0000160  phosphorelay signal transduction system\n",
      "BIOLOGICAL_PROCESS  GO:0000162  tryptophan biosynthetic process\n",
      "BIOLOGICAL_PROCESS  GO:0000256  allantoin catabolic process\n",
      "BIOLOGICAL_PROCESS  GO:0000271  polysaccharide biosynthetic process\n",
      "BIOLOGICAL_PROCESS  GO:0000272  polysaccharide catabolic process\n",
      "BIOLOGICAL_PROCESS  GO:0000413  protein peptidyl-prolyl isomerization\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Empty file\n",
    "printf \"\" > GO_terms_tmp.tsv\n",
    "\n",
    "for XML in InterproResults/*xml; do\n",
    "\n",
    "    # Retrieve go terms; remove attribute names; remove leading spaces and \n",
    "    # replace the first two blank spaces by tabs\n",
    "    PRE_CLEANED_GO=$(grep -E \"<go-xref\\s+category\" \"${XML}\" |\n",
    "                         sed -r 's/(<go-xref category=|db=\"GO\"\\s+id=|name=|\\/>|\")//g' |\n",
    "                         sed -r 's/^\\s+//g' |\n",
    "                         sed -r 's/\\s+/\\t/1' |\n",
    "                         sed -r 's/\\s+/\\t/2')\n",
    "    \n",
    "    echo \"${PRE_CLEANED_GO}\" >> GO_terms_tmp.tsv\n",
    "\n",
    "done\n",
    "\n",
    "# Removing go terms repeated\n",
    "sort GO_terms_tmp.tsv | uniq > GO_terms.tsv && rm GO_terms_tmp.tsv\n",
    "\n",
    "# Adding NaN at the end and headers at the beginning\n",
    "printf \"NaN\\tNaN\\tNaN\\n\" >> GO_terms.tsv\n",
    "sed -i \"1s/^/category\\tid_go\\tdescription_go\\n/\" GO_terms.tsv\n",
    "\n",
    "column -s$'\\t' -t GO_terms.tsv | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6859dda-f868-4c29-a5ff-f61838a88538",
   "metadata": {
    "tags": []
   },
   "source": [
    "As you can see, the new file named *GO_terms.tsv* has three columns: first one with the category of the GO term, a second one with the GO id and the last one with the description about that term.\n",
    "\n",
    "Now it's time to process the remained labels. Because of the complexity of the XML files (e.g many nested tags) the use of libraries such as ```xml.etree.ElementTree``` in pyhon does not seem appropiate in terms of for loops needed and maybe in terms of legibility (but I'll use regex haha), same case for trying to use ```pandas``` so again, I'll do that using bash, feel free to find another way to acomplish this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "436abd88-5288-40f7-8bc9-61cf4b94a4a5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id_protein\tid_analysis\ttype\n",
      "NP_416254.1\tG3DSA:3.40.50.620:FF:000015\tNULL_TYPE\n",
      "NP_416254.1\tPF02540\tDOMAIN\n",
      "NP_416254.1\tPTHR23090\tFAMILY\n",
      "NP_416254.1\tcd00553\tFAMILY\n",
      "NP_416254.1\tSSF52402\tNULL_TYPE\n",
      "NP_416630.1\tPF07694\tDOMAIN\n",
      "NP_416630.1\tPF06580\tDOMAIN\n",
      "NP_416630.1\tG3DSA:3.30.450.40:FF:000013\tNULL_TYPE\n",
      "NP_416630.1\tPTHR34220\tNULL_TYPE\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for XML in InterproResults/*xml; do\n",
    "\n",
    "    OUTPUT_NAME=$(basename \"${XML}\" | sed -r 's/$/\\.tsv/g')\n",
    "\n",
    "    # Retrieve all tags and attributes; retrieve only important attributes; clean attribute names \n",
    "    # and quote \"\n",
    "    PRE_CLEANED_XML=$(grep -P \"(<xref |<signature |<entry ac|<protein>)\" \"${XML}\" |\n",
    "                          grep -Po '(<protein>|<xref id=\"[^\"]+\"|<signature ac=\"[^\"]+\"|type=\"[^\"]+\")' |\n",
    "                          sed -r 's/(^<\\w+ |\")//g')\n",
    "\n",
    "    # merge multiples group of proteins into a single row separated by |;\n",
    "    # merge type with method ID into a single row separated by tab;\n",
    "    # clean attributes names\n",
    "    PRE_FORMATED_XML=$(echo \"${PRE_CLEANED_XML}\" |\n",
    "                           sed ':r;$!{N;br};s/\\nid/\\|id/g' |\n",
    "                           sed ':r;$!{N;br};s/\\ntype/\\ttype/g' |\n",
    "                           sed -r 's/<protein>\\|//g' |\n",
    "                           sed -r 's/ac=|type=//g')\n",
    "\n",
    "    # Parser protein ID (clear format) and get tabular form\n",
    "    echo \"${PRE_FORMATED_XML}\" |\n",
    "        perl -ne 'if($_ =~ /id=/){\n",
    "                        chomp($_);\n",
    "                        $header= ($_ =~ s/id=//gr);\n",
    "                } else{\n",
    "                        chomp($_);\n",
    "                        $new_string=$header . \"\\t\" . $_;\n",
    "                        $len = 3 - scalar(split(\"\\t\", $new_string));\n",
    "                        print $new_string,\"\\tNULL_TYPE\"x$len,\"\\n\"\n",
    "                }' > InterproParsedXML/$OUTPUT_NAME\n",
    "    \n",
    "    # Adding header\n",
    "    sed -i '1s/^/id_protein\\tid_analysis\\ttype\\n/' InterproParsedXML/$OUTPUT_NAME\n",
    "\n",
    "done\n",
    "\n",
    "# Ecoli file\n",
    "head InterproParsedXML/GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fc4bd4d-2daa-4b59-8985-41007dc48b89",
   "metadata": {},
   "source": [
    "From this file, we can load and map by protein ID and Method ID the XML file parsed and the tsv from interproscan for every organism, we'll take _E.coli_ as an example of the logic to follow:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "dd62775f-07e6-45fd-85e1-8d60ed6bb8f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Reading two into two dataFrames the tsv file from interproscan and the tsv file obtained from the \n",
    "# processing of the xml file\n",
    "\n",
    "path_interpro_parsed = \"InterproParsedXML/\"\n",
    "\n",
    "xml_parsed_df = pd.read_csv(os.path.join(path_interpro_parsed,\"GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv\"),\n",
    "                            sep=\"\\t\")\n",
    "\n",
    "tsv_df = pd.read_csv(os.path.join(path_interpro_results, \"GCF_000005845.2_E_coli_K12_genomic.faa.tsv\"),\n",
    "                     sep=\"\\t\", )\n",
    "\n",
    "# Replacing null values (NULL_type and score - in both df) by np.NaN\n",
    "for col in xml_parsed_df:\n",
    "    xml_parsed_df[col] = xml_parsed_df[col].replace(\"NULL_TYPE\", np.NaN)\n",
    "\n",
    "for col in tsv_df:\n",
    "    tsv_df[col] = tsv_df[col].replace(\"-\", np.NaN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "5d6f19f2-d42b-4ef0-9ea3-203a0b53d9ab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 20195 entries, 0 to 20194\n",
      "Data columns (total 3 columns):\n",
      " #   Column       Non-Null Count  Dtype \n",
      "---  ------       --------------  ----- \n",
      " 0   id_protein   20195 non-null  object\n",
      " 1   id_analysis  20195 non-null  object\n",
      " 2   type         11561 non-null  object\n",
      "dtypes: object(3)\n",
      "memory usage: 473.4+ KB\n"
     ]
    }
   ],
   "source": [
    "xml_parsed_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "c4558211-002f-4435-a223-7b34efc1bdcf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 20615 entries, 0 to 20614\n",
      "Data columns (total 14 columns):\n",
      " #   Column         Non-Null Count  Dtype  \n",
      "---  ------         --------------  -----  \n",
      " 0   id_protein     20615 non-null  object \n",
      " 1   md5            20615 non-null  object \n",
      " 2   seq_len        20615 non-null  int64  \n",
      " 3   analysis       20615 non-null  object \n",
      " 4   id_analysis    20615 non-null  object \n",
      " 5   desc_analysis  20529 non-null  object \n",
      " 6   start          20615 non-null  int64  \n",
      " 7   stop           20615 non-null  int64  \n",
      " 8   score          20615 non-null  float64\n",
      " 9   status         20615 non-null  object \n",
      " 10  date           20615 non-null  object \n",
      " 11  id_inter       11921 non-null  object \n",
      " 12  desc_inter     11921 non-null  object \n",
      " 13  go             5540 non-null   object \n",
      "dtypes: float64(1), int64(3), object(10)\n",
      "memory usage: 2.2+ MB\n"
     ]
    }
   ],
   "source": [
    "tsv_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b62169-e45d-4403-a5eb-ce7225e576d0",
   "metadata": {},
   "source": [
    "From the first file and the second one respectively, both *go* and *id_protein* columns have multiples values in a single row, in order to connect the table we are creating with the **GO_terms.tsv** table for each protein, we'll convert the DataFrame in a \"long\" format dataframe expanding the multiples values found in *go* and *id_protein* into multiples rows describing the same result but, this time, each row will have only one GO term and id as exemplified below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "c2e9e052-e706-4f55-992b-41762f40b0bc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GO:1|GO:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2|3</td>\n",
       "      <td>GO:3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ids         go\n",
       "0    0  GO:1|GO:2\n",
       "1    1        NaN\n",
       "2  2|3       GO:3"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids = [0,1,\"2|3\"]\n",
    "go_terms = [\"GO:1|GO:2\",np.NaN,\"GO:3\"]\n",
    "\n",
    "df_shown = pd.DataFrame({\"ids\":ids, \"go\":go_terms})\n",
    "df_shown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "52bfd637-58e4-4562-b411-ae7ddcd8184c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GO:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>GO:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>GO:3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>GO:3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ids    go\n",
       "0    0  GO:1\n",
       "0    0  GO:2\n",
       "1    1   NaN\n",
       "2    2  GO:3\n",
       "2    3  GO:3"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ids = np.array([0,0,1,2,3])\n",
    "go_terms = [\"GO:1\",\"GO:2\", np.NaN, \"GO:3\", \"GO:3\"]\n",
    "\n",
    "df_expected = pd.DataFrame({\"ids\":ids, \"go\":go_terms})\n",
    "df_expected.index = [0,0,1,2,2]\n",
    "df_expected"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c42b99-8a5c-40fc-a6f6-f8cff763ef16",
   "metadata": {},
   "source": [
    "Note that maintaining the index is important since it will be used to map with the original table. Let's do it using only the *go* column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "15d0ffaa-0602-4589-9a82-026b20c4b0b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Convert the GO terms found into a dataframe of multiples columns depending of the number of GO terms\n",
    "go_tsv = tsv_df[\"go\"].str.split(\"|\").apply(pd.Series)\n",
    "\n",
    "# \"Merge\" all the columns from the same index into multiples rows of one colum\n",
    "go_tsv = go_tsv.copy().T\n",
    "go_tsv_long = go_tsv.melt(var_name=\"index_tmp\", value_name=\"id_go\").drop_duplicates()\n",
    "\n",
    "# Masking to remove duplicated index with np.NaN value\n",
    "mask = ~(go_tsv_long[\"index_tmp\"].duplicated(keep=False) & go_tsv_long[\"id_go\"].isna())\n",
    "go_tsv_long = go_tsv_long[mask].copy().set_index(\"index_tmp\")\n",
    "\n",
    "# Left join with the original table\n",
    "tsv_df = tsv_df.merge(go_tsv_long, how=\"left\", right_index=True, left_index=True).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "05d7a87a-c162-4d39-94d8-7639a5a155f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_protein</th>\n",
       "      <th>md5</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>analysis</th>\n",
       "      <th>id_analysis</th>\n",
       "      <th>desc_analysis</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>id_inter</th>\n",
       "      <th>desc_inter</th>\n",
       "      <th>go</th>\n",
       "      <th>id_go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0004359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0003952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0004359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR23090</td>\n",
       "      <td>NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE</td>\n",
       "      <td>11</td>\n",
       "      <td>242</td>\n",
       "      <td>7.600000e-32</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952|GO:0004359|GO:0005737|GO:0009435</td>\n",
       "      <td>GO:0009435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>FunFam</td>\n",
       "      <td>G3DSA:3.40.50.620:FF:000015</td>\n",
       "      <td>NH(3)-dependent NAD(+) synthetase</td>\n",
       "      <td>1</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>SSF52402</td>\n",
       "      <td>Adenine nucleotide alpha hydrolases-like</td>\n",
       "      <td>3</td>\n",
       "      <td>273</td>\n",
       "      <td>1.450000e-86</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id_protein                               md5  seq_len     analysis  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275          CDD   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275          CDD   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275          CDD   \n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275          CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      PANTHER   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      PANTHER   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      PANTHER   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      PANTHER   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275       FunFam   \n",
       "3  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275  SUPERFAMILY   \n",
       "\n",
       "                   id_analysis                                desc_analysis  \\\n",
       "0                      cd00553                                 NAD_synthase   \n",
       "0                      cd00553                                 NAD_synthase   \n",
       "0                      cd00553                                 NAD_synthase   \n",
       "0                      cd00553                                 NAD_synthase   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "1                    PTHR23090  NH 3 /GLUTAMINE-DEPENDENT NAD +  SYNTHETASE   \n",
       "2  G3DSA:3.40.50.620:FF:000015            NH(3)-dependent NAD(+) synthetase   \n",
       "3                     SSF52402     Adenine nucleotide alpha hydrolases-like   \n",
       "\n",
       "   start  stop         score status        date   id_inter         desc_inter  \\\n",
       "0     17   262  1.857550e-84      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "0     17   262  1.857550e-84      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "0     17   262  1.857550e-84      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "0     17   262  1.857550e-84      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1     11   242  7.600000e-32      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1     11   242  7.600000e-32      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1     11   242  7.600000e-32      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "1     11   242  7.600000e-32      T  08-02-2023  IPR003694  NAD(+) synthetase   \n",
       "2      1   274  0.000000e+00      T  08-02-2023        NaN                NaN   \n",
       "3      3   273  1.450000e-86      T  08-02-2023        NaN                NaN   \n",
       "\n",
       "                                            go       id_go  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0003952  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0004359  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0005737  \n",
       "0  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0009435  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0003952  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0004359  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0005737  \n",
       "1  GO:0003952|GO:0004359|GO:0005737|GO:0009435  GO:0009435  \n",
       "2                                          NaN         NaN  \n",
       "3                                          NaN         NaN  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a3a7633-5f09-426e-8517-46af22600c4a",
   "metadata": {},
   "source": [
    "We have split correctly the *go* column! Now we don't need that column anymore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "56ac82f2-c763-4fce-a2ad-1e4260ddafec",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 25644 entries, 0 to 20614\n",
      "Data columns (total 14 columns):\n",
      " #   Column         Non-Null Count  Dtype  \n",
      "---  ------         --------------  -----  \n",
      " 0   id_protein     25644 non-null  object \n",
      " 1   md5            25644 non-null  object \n",
      " 2   seq_len        25644 non-null  int64  \n",
      " 3   analysis       25644 non-null  object \n",
      " 4   id_analysis    25644 non-null  object \n",
      " 5   desc_analysis  25522 non-null  object \n",
      " 6   start          25644 non-null  int64  \n",
      " 7   stop           25644 non-null  int64  \n",
      " 8   score          25644 non-null  float64\n",
      " 9   status         25644 non-null  object \n",
      " 10  date           25644 non-null  object \n",
      " 11  id_inter       16950 non-null  object \n",
      " 12  desc_inter     16950 non-null  object \n",
      " 13  id_go          10569 non-null  object \n",
      "dtypes: float64(1), int64(3), object(10)\n",
      "memory usage: 2.9+ MB\n"
     ]
    }
   ],
   "source": [
    "tsv_df.drop(\"go\", axis=1, inplace=True)\n",
    "tsv_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57724b23-a440-4590-919e-fdc2ba12ea65",
   "metadata": {},
   "source": [
    "Now we would have to do the same with the *id_protein* column of the parsed xml file and we will do it, but for now I just want to continue explaining what would be the next thing to do and in this way we would have to join both tables from both files. To do this, we will make a shared column that will serve as a key to join both datasets based on the id of the protein and of the method used, columns repeated will be deleted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "96e544bc-8538-41a2-ab1c-4df202d32ecf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_protein_x</th>\n",
       "      <th>md5</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>analysis</th>\n",
       "      <th>id_analysis_x</th>\n",
       "      <th>desc_analysis</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>id_inter</th>\n",
       "      <th>desc_inter</th>\n",
       "      <th>id_go</th>\n",
       "      <th>shared_index</th>\n",
       "      <th>id_protein_y</th>\n",
       "      <th>id_analysis_y</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952</td>\n",
       "      <td>NP_416254.1\\tcd00553</td>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0004359</td>\n",
       "      <td>NP_416254.1\\tcd00553</td>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>NP_416254.1\\tcd00553</td>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id_protein_x                               md5  seq_len analysis  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      CDD   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      CDD   \n",
       "\n",
       "  id_analysis_x desc_analysis  start  stop         score status        date  \\\n",
       "0       cd00553  NAD_synthase     17   262  1.857550e-84      T  08-02-2023   \n",
       "1       cd00553  NAD_synthase     17   262  1.857550e-84      T  08-02-2023   \n",
       "2       cd00553  NAD_synthase     17   262  1.857550e-84      T  08-02-2023   \n",
       "\n",
       "    id_inter         desc_inter       id_go          shared_index  \\\n",
       "0  IPR003694  NAD(+) synthetase  GO:0003952  NP_416254.1\\tcd00553   \n",
       "1  IPR003694  NAD(+) synthetase  GO:0004359  NP_416254.1\\tcd00553   \n",
       "2  IPR003694  NAD(+) synthetase  GO:0005737  NP_416254.1\\tcd00553   \n",
       "\n",
       "  id_protein_y id_analysis_y    type  \n",
       "0  NP_416254.1       cd00553  FAMILY  \n",
       "1  NP_416254.1       cd00553  FAMILY  \n",
       "2  NP_416254.1       cd00553  FAMILY  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# shared_index column will be used as key in both files\n",
    "tsv_df[\"shared_index\"] = tsv_df[\"id_protein\"] + \"\\t\" + tsv_df[\"id_analysis\"]\n",
    "xml_parsed_df[\"shared_index\"] = xml_parsed_df[\"id_protein\"] + \"\\t\" + xml_parsed_df[\"id_analysis\"]\n",
    "\n",
    "# Left join\n",
    "full_tsv = tsv_df.merge(xml_parsed_df, on=\"shared_index\", how=\"left\")\n",
    "full_tsv.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "658226c1-8b7d-419c-a098-0ed5964390f7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_protein</th>\n",
       "      <th>md5</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>analysis</th>\n",
       "      <th>id_analysis_x</th>\n",
       "      <th>desc_analysis</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>score</th>\n",
       "      <th>status</th>\n",
       "      <th>date</th>\n",
       "      <th>id_inter</th>\n",
       "      <th>desc_inter</th>\n",
       "      <th>id_go</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0003952</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0004359</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_416254.1</td>\n",
       "      <td>ea37cb76b651f90db9b8d4dbf6861145</td>\n",
       "      <td>275</td>\n",
       "      <td>CDD</td>\n",
       "      <td>cd00553</td>\n",
       "      <td>NAD_synthase</td>\n",
       "      <td>17</td>\n",
       "      <td>262</td>\n",
       "      <td>1.857550e-84</td>\n",
       "      <td>T</td>\n",
       "      <td>08-02-2023</td>\n",
       "      <td>IPR003694</td>\n",
       "      <td>NAD(+) synthetase</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>FAMILY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    id_protein                               md5  seq_len analysis  \\\n",
       "0  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      CDD   \n",
       "1  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      CDD   \n",
       "2  NP_416254.1  ea37cb76b651f90db9b8d4dbf6861145      275      CDD   \n",
       "\n",
       "  id_analysis_x desc_analysis  start  stop         score status        date  \\\n",
       "0       cd00553  NAD_synthase     17   262  1.857550e-84      T  08-02-2023   \n",
       "1       cd00553  NAD_synthase     17   262  1.857550e-84      T  08-02-2023   \n",
       "2       cd00553  NAD_synthase     17   262  1.857550e-84      T  08-02-2023   \n",
       "\n",
       "    id_inter         desc_inter       id_go    type  \n",
       "0  IPR003694  NAD(+) synthetase  GO:0003952  FAMILY  \n",
       "1  IPR003694  NAD(+) synthetase  GO:0004359  FAMILY  \n",
       "2  IPR003694  NAD(+) synthetase  GO:0005737  FAMILY  "
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_tsv.drop([\"shared_index\",\"id_protein_y\",\"id_analysis_y\"], axis=1, inplace=True)\n",
    "full_tsv.rename({\"id_protein_x\":\"id_protein\"}, inplace=True, axis=1)\n",
    "full_tsv.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74694ea8-50f2-458b-8753-4aef3a1d5706",
   "metadata": {},
   "source": [
    "Remember we don't split the *id_protein* column and the DataFrame is not entirely valid.\n",
    "\n",
    "Now we need to do the same for every organism and split *id_protein*, I'll create some functions of the codes shown above, so they are going to do all the hard work. In addition, a short code will be added to save into a file the tables created and this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "62332ac0-21fd-4ec8-8df7-4fcff5537067",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def LoadFiles(file_1:str, file_2:str, keydict:dict=None) -> (pd.DataFrame, pd.DataFrame):\n",
    "    \n",
    "    # This function read a tsv file and a parsed xml file obtained from interproscan into a pandas DataFrame\n",
    "    \n",
    "    print(f\"Reading files into DataFrames: {file_1}\")\n",
    "    print(f\"                               {file_2}\")\n",
    "    \n",
    "    df_1 = pd.read_csv(file_1, **keydict)\n",
    "    df_2 = pd.read_csv(file_2, **keydict)\n",
    "    \n",
    "    return df_1, df_2\n",
    "\n",
    "def ReplaceInFrame(data:pd.DataFrame, to_replace:str, by=np.NaN) -> pd.DataFrame:\n",
    "    \n",
    "    # This function replace a value for another in a dataframe\n",
    "    print(f\"Replacing {to_replace} by {by}\")\n",
    "    \n",
    "    for col in data:\n",
    "        data[col] = data[col].replace(to_replace, by)\n",
    "    \n",
    "    return data\n",
    "\n",
    "def ParseMultiValuesFrames(data:pd.DataFrame, col_target, sep_target,\n",
    "                           value_name:str, keydict:dict) -> pd.DataFrame:\n",
    "    \n",
    "    # This function split multiples values found in a column and row of a dataFrame\n",
    "    # into multiples lines containing only one value term in a dataframe  \n",
    "    \n",
    "    print(f\"Parsing multivalues in column {col_target}\")\n",
    "    \n",
    "    df_splitted = data[col_target].str.split(sep_target).apply(pd.Series)\n",
    "\n",
    "    # \"Merge\" all the columns from the same index into multiples rows of one colum\n",
    "    df_splitted_T = df_splitted.copy().T\n",
    "    df_splitted_T = df_splitted_T.melt(var_name=\"index_tmp\", value_name=\"column_tmp\").drop_duplicates()\n",
    "\n",
    "    # Masking to remove duplicated index with np.NaN value\n",
    "    mask = ~(df_splitted_T[\"index_tmp\"].duplicated(keep=False) & df_splitted_T[\"column_tmp\"].isna())\n",
    "    df = df_splitted_T[mask].copy().set_index(\"index_tmp\")\n",
    "    full_df = data.merge(df, **keydict)\n",
    "    full_df = full_df.drop(col_target, axis=1)\n",
    "    \n",
    "    return full_df.rename({\"column_tmp\": value_name}, axis=1)\n",
    "    \n",
    "def CreateSharedIndex(df_1:pd.DataFrame, df_1_cols:list,\n",
    "                      df_2:pd.DataFrame, df_2_cols:list,\n",
    "                      shared_index:str, sep=\"\\t\") -> (pd.DataFrame, pd.DataFrame):\n",
    "    \n",
    "    # This function create a shared index between two data frames\n",
    "    \n",
    "    print(f\"Creating shared_index = {shared_index}\")\n",
    "    \n",
    "    df_1[shared_index] = df_1[df_1_cols[0]] + sep + df_1[df_1_cols[0]]\n",
    "    df_2[shared_index] = df_2[df_2_cols[0]] + sep + df_2[df_2_cols[1]]\n",
    "    \n",
    "    return df_1, df_2\n",
    "\n",
    "def JoinFrames(left:pd.DataFrame, right:pd.DataFrame, keydict:dict, drop:list=None) -> pd.DataFrame:\n",
    "    \n",
    "    # This function do a join operation between two dataframes\n",
    "    \n",
    "    print(f\"Joining frames with parameters: {keydict}\")\n",
    "    df = left.merge(right, **keydict)\n",
    "    \n",
    "    if drop:\n",
    "        df.drop(drop, axis=1, inplace=True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6442ecc-f295-4097-844d-a1ab4e370458",
   "metadata": {},
   "source": [
    "TSV file are already loaded, we'll need to load now the created xml.tsv files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "fcc1464a-4dc5-4dbf-b1df-424bacd07e36",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv',\n",
       " 'GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.xml.tsv',\n",
       " 'GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.xml.tsv',\n",
       " 'GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.xml.tsv',\n",
       " 'GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.xml.tsv',\n",
       " 'GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.xml.tsv']"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_files_names_parsed = os.listdir(path_interpro_parsed)\n",
    "\n",
    "# Filter only the files ending with \"xml.tsv\"\n",
    "xml_tsv_files_names = [f for f in all_files_names_parsed if f.endswith(\"xml.tsv\") ]\n",
    "xml_tsv_files_names.sort()\n",
    "xml_tsv_files_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "83a97453-005e-435c-98ee-bb1eaf5d4848",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000005845.2_E_coli_K12_genomic.faa.tsv',\n",
       " 'GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv',\n",
       " 'GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv',\n",
       " 'GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv',\n",
       " 'GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv',\n",
       " 'GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv']"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_files_names_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "e4621950-105e-4e90-8d76-552639b03d3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading files into DataFrames: InterproParsedXML/GCF_000005845.2_E_coli_K12_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000005845.2_E_coli_K12_genomic.faa.tsv\n",
      "Replacing NULL_TYPE by nan\n",
      "Replacing - by nan\n",
      "Parsing multivalues in column id_protein\n",
      "Parsing multivalues in column go\n",
      "Creating shared_index = shared_index\n",
      "Joining frames with parameters: {'on': 'shared_index', 'how': 'left'}\n",
      "Saving frame into InterproFullMerge/GCF_000005845.2_E_coli_K12_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv\n",
      "Replacing NULL_TYPE by nan\n",
      "Replacing - by nan\n",
      "Parsing multivalues in column id_protein\n",
      "Parsing multivalues in column go\n",
      "Creating shared_index = shared_index\n",
      "Joining frames with parameters: {'on': 'shared_index', 'how': 'left'}\n",
      "Saving frame into InterproFullMerge/GCF_000006765.1_ASM676v1_P_aeruginosa_PA01_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv\n",
      "Replacing NULL_TYPE by nan\n",
      "Replacing - by nan\n",
      "Parsing multivalues in column id_protein\n",
      "Parsing multivalues in column go\n",
      "Creating shared_index = shared_index\n",
      "Joining frames with parameters: {'on': 'shared_index', 'how': 'left'}\n",
      "Saving frame into InterproFullMerge/GCF_000006945.2_ASM694v2_S_enterica_LT2_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv\n",
      "Replacing NULL_TYPE by nan\n",
      "Replacing - by nan\n",
      "Parsing multivalues in column id_protein\n",
      "Parsing multivalues in column go\n",
      "Creating shared_index = shared_index\n",
      "Joining frames with parameters: {'on': 'shared_index', 'how': 'left'}\n",
      "Saving frame into InterproFullMerge/GCF_000009045.1_ASM904v1_B_subtilis_168_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv\n",
      "Replacing NULL_TYPE by nan\n",
      "Replacing - by nan\n",
      "Parsing multivalues in column id_protein\n",
      "Parsing multivalues in column go\n",
      "Creating shared_index = shared_index\n",
      "Joining frames with parameters: {'on': 'shared_index', 'how': 'left'}\n",
      "Saving frame into InterproFullMerge/GCF_000009645.1_ASM964v1_S_aureus_N315_genomic.faa.tsv.full.tsv\n",
      "\n",
      "Reading files into DataFrames: InterproParsedXML/GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.xml.tsv\n",
      "                               InterproResults/GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv\n",
      "Replacing NULL_TYPE by nan\n",
      "Replacing - by nan\n",
      "Parsing multivalues in column id_protein\n",
      "Parsing multivalues in column go\n",
      "Creating shared_index = shared_index\n",
      "Joining frames with parameters: {'on': 'shared_index', 'how': 'left'}\n",
      "Saving frame into InterproFullMerge/GCF_000195955.2_ASM19595v2_M_tuberculosis_H37Rv_genomic.faa.tsv.full.tsv\n",
      "\n",
      "All done\n"
     ]
    }
   ],
   "source": [
    "# Arguments for the functions and folder to save files\n",
    "output_dir = \"InterproFullMerge\"\n",
    "arguments_load_function = {\"sep\":\"\\t\"}\n",
    "arguments_parse_function = {\"how\":\"left\", \"right_index\":True, \"left_index\":True}\n",
    "arguments_join_function = {\"on\":\"shared_index\", \"how\":\"left\"}\n",
    "\n",
    "# Doing the same as shown above but using all organisms\n",
    "for t_f, x_f in zip(tsv_files_names_results, xml_tsv_files_names):\n",
    "    \n",
    "    tsv_path = os.path.join(path_interpro_results, t_f)\n",
    "    xml_path = os.path.join(path_interpro_parsed, x_f)\n",
    "    xml_df, tsv_df = LoadFiles(xml_path, tsv_path, arguments_load_function)\n",
    "    \n",
    "    xml_df = ReplaceInFrame(xml_df, to_replace=\"NULL_TYPE\")\n",
    "    tsv_df = ReplaceInFrame(tsv_df, to_replace=\"-\")\n",
    "    \n",
    "    xml_df = ParseMultiValuesFrames(xml_df, col_target=\"id_protein\",\n",
    "                                    sep_target=\"|\", value_name=\"id_protein\",\n",
    "                                    keydict=arguments_parse_function)\n",
    "    \n",
    "    tsv_df = ParseMultiValuesFrames(tsv_df, col_target=\"go\", \n",
    "                                    sep_target=\"|\", value_name=\"id_go\",\n",
    "                                    keydict=arguments_parse_function)\n",
    "    \n",
    "    xml_df, tsv_df = CreateSharedIndex(df_1=xml_df, df_1_cols=[\"id_protein\", \"id_analysis\"],\n",
    "                                       df_2=tsv_df, df_2_cols=[\"id_protein\", \"id_analysis\"],\n",
    "                                       shared_index=\"shared_index\")\n",
    "    \n",
    "    full_df = JoinFrames(tsv_df, xml_df,\n",
    "                         arguments_join_function, [\"shared_index\",\"id_protein_y\",\"id_analysis_y\"])\n",
    "    \n",
    "    full_df.rename({\"id_protein_x\":\"id_protein\",\"id_analysis_x\":\"id_analysis\"},\n",
    "                   inplace=True, axis=1)\n",
    "    \n",
    "    # Saving frame into a file ending with \".full\" \n",
    "    output_file_path = os.path.join(output_dir, t_f + \".full.tsv\")\n",
    "    print(f\"Saving frame into {output_file_path}\\n\")\n",
    "    full_df.to_csv(output_file_path, sep=\"\\t\", header=True, index=False, na_rep=\"NaN\")\n",
    "\n",
    "print(\"All done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73763e55-8e8f-490d-bee3-2f601d44a2ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
